### PR TITLE
Move reference-holding last_snapshot_storages from ABS to AHV

### PR DESCRIPTION
**Pull request by xiangzhu70**
#### Problem
Constructing a bank should be done from a POST snapshot (to have the calculated hash).  So the reference counts should be held for the latest POST snapshot dir, not just the latest dir.  AHV is the right place, because it converts PRE to POST.

#### Summary of Changes
Move last_snapshot_storage from ABS to AHV
Remove storages cloning inside ABS, because there is no need to return a copy from handle_snapshot_request
There is a new cloning in AHV at the start of AHV handling a new accounts_package.  The total number of cloning is not changed.

Fixes #

### CODE REVIEW

File: core/src/accounts_hash_verifier.rs [hunk 1]
```diff
@@ -80,6 +84,7 @@ impl AccountsHashVerifier {
                         continue;
                     };
                     info!("handling accounts package: {accounts_package:?}");
+                    let snapshot_storages = accounts_package.snapshot_storages.clone();
```
[the remark this answers is not in the document]
Comment (xiangzhu70) [reply core/src/accounts_hash_verifier.rs]: Got it.  Done.

File: core/src/accounts_hash_verifier.rs [hunk 2]
```diff
@@ -109,7 +115,28 @@ impl AccountsHashVerifier {
...
                         ("handling-time-us", handling_time_us, i64),
                     );
+
+                    // Done processing the current snapshot, so the current snapshot dir
+                    // has been converted to POST state.  It is the time to update
+                    // last_snapshot_storages to release the reference counts for the
+                    // previous POST snapshot dir, and save the new ones for the new
+                    // POST snapshot dir.
+                    last_snapshot_storages = Some(snapshot_storages);
+                    debug!(
```
Comment (apfitzge) [inline core/src/accounts_hash_verifier.rs:125]: I see they were just moved, but I'm curious about why this debug message is different from the one on line 133.

I think if we want to make the logging different so we can determine which is from a **new** snapshot, then we should include that in the log message.

(feel free to do this in a separate PR - it's not related to the move of this code)

**Review by apfitzge — APPROVED** [review]
Code change looks good to me, because we want to keep the last POST snapshot AHV seems like the right place to keep references to me.

Added a comment about debug messages as a follow-up.

File: core/src/accounts_hash_verifier.rs (hunk 2 above)
Comment (xiangzhu70) [reply core/src/accounts_hash_verifier.rs:125]: I copied them from the same positions in ABS, one in the loop, one out of the loop.  The in-loop and out-loop messages may not be the same.

Comment (xiangzhu70) [reply core/src/accounts_hash_verifier.rs:125]: Ok, these two messages mean the same thing.  I will make them the same.

Comment (xiangzhu70) [reply core/src/accounts_hash_verifier.rs]: updated it.

File: core/src/accounts_hash_verifier.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (xiangzhu70) [reply core/src/accounts_hash_verifier.rs]: updated it.

File: core/src/accounts_hash_verifier.rs [hunk 3]
```diff
@@ -109,7 +115,28 @@ impl AccountsHashVerifier {
...
+                        "Number of snapshot storages kept alive for fastboot: {}",
+                        last_snapshot_storages
+                            .as_ref()
+                            .map(|storages| storages.len())
+                            .unwrap_or(0)
+                    );
```
[the remark this answers is not in the document]
Comment (xiangzhu70) [reply core/src/accounts_hash_verifier.rs]: updated it.

File: core/src/accounts_hash_verifier.rs [hunk 4]
```diff
@@ -63,6 +63,10 @@ impl AccountsHashVerifier {
             .spawn(move || {
                 info!("AccountsHashVerifier has started");
                 let mut hashes = vec![];
+                // To support fastboot, we must ensure the storages used in the latest POST snapshot are
+                // not recycled nor removed early.  Hold an Arc of their AppendVecs to prevent them from
+                // expiring.
+                let mut last_snapshot_storages: Option<Vec<Arc<AccountStorageEntry>>> = None;
```
Comment (apfitzge) [inline core/src/accounts_hash_verifier.rs:69]: Maybe we landed on this in a previous PR, but the type annotation here is not required.

Comment (xiangzhu70) [reply core/src/accounts_hash_verifier.rs]: Ok, removed the type annotation

**Review by apfitzge — APPROVED** [review]
lgtm - just wait for the CI first

**Pull request merged**